### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by LoxiasCam**
# Railsy Forms

For this project, I created several forms using Rails and HTML according to the Forms assignment.

## User model created with validations:

  *  username
  *  password
  *  email

## Views created with forms:

*    Create
 *   New
  *  Update
  *  Show
  * Delete

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad**. 

### Well-done 💯 
    1. The linter checks run just fine.
    2. The readme file is clear and concise.

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [ ] Check the comments below for sections to improve

#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
+  root to: 'users#index'
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :users, only: [:new, :create, :show, :edit, :update]
```
Comment (mke2111) [inline config/routes.rb]: - [x] In `config/routes.rb`, these are the only actions you need to handle. `[:new, :create, :edit, :update]`

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    @user = User.new
+  end
+
+  def create
+    # @user = User.new(username: params[:username], email: params[:email], password: params[:password])
+    @user = User.new(user_params)
+
+    if @user.save
+      redirect_to root_path
+    else
```
Comment (mke2111) [inline app/controllers/users_controller.rb:20]: Once a new user is created successfully, you should redirect your app to the `new_user_path`

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
+    if @user.update(user_params)
+      flash.notice = "User '#{@user.username}' edited!"
+      redirect_to @user
```
Comment (mke2111) [inline app/controllers/users_controller.rb]: If this action is successful as well, redirect to `edit_user_path(<USER_ID>)`.

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+  validates :username, presence: true
+  validates :password, presence: true
+  validates :email, presence: true
+end
```
Comment (mke2111) [inline app/models/user.rb:3]: OPTIONAL
- Always keep your code DRY. You can simply write the three lines like this
`validates: :username, :password, :email, presence: true`

File: app/views/users/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<!-- <form action="/users" method="post" accept-charset="UTF-8">
+  <input type="hidden" name="authenticity_token" value="<%= form_authenticity_token %>">
+  <br><br>
+  <label for="username" >Username:</label>
```
Comment (mke2111) [inline app/views/users/new.html.erb:4]: This `form` hasn't been commented out entirely.

Use this syntax instead

```
<%=
begin
%>

# Code 

<%=
end
%>
```

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+<!-- <%= form_tag("/users", method: "post") do %>
+  <br><br>
+  <%= label_tag(:username, "Username:") %>
+  <%= text_field_tag(:username) %><br><br>
+  <%= label_tag(:email, "Email:") %>
+  <%= text_field_tag(:email) %><br><br>
+  <%= label_tag(:password, "Password:") %>
+  <%= password_field_tag(:password) %><br><br>
+  <%= submit_tag "Submit" %>
+<% end %> -->
```
Comment (mke2111) [inline app/views/users/new.html.erb]: Use the above syntax to comment out this `form_tag` successfully

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  <label for="password">Password:</label>
+  <input type="password" id="password" name="user[password]"><br><br>
+  <input type="submit" value="Submit">
+</form> -->
+<!-- <%= form_tag("/users", method: "post") do %>
+  <br><br>
+  <%= label_tag(:username, "Username:") %>
+  <%= text_field_tag(:username) %><br><br>
+  <%= label_tag(:email, "Email:") %>
+  <%= text_field_tag(:email) %><br><br>
```
Comment (mke2111) [inline app/views/users/new.html.erb]: Change this to `email_field_tag`

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  <%= label_tag(:password, "Password:") %>
+  <%= password_field_tag(:password) %><br><br>
+  <%= submit_tag "Submit" %>
+<% end %> -->
+<!-- <%= form_with model: @user, method: :post do |form| %>
+  <br><br>
+  <%= form.label :username, "Username:" %>
+  <%= form.text_field :username, :value => 'Please enter your username here'%><br><br>
+  <%= form.label :email, "Email:" %>
+  <%= form.text_field :email, :value => 'user@domain.com' %><br><br>
```
Comment (mke2111) [inline app/views/users/new.html.erb]: Change this to `form.email_field`

File: app/views/users/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  <%= form.label :password, "Password:" %>
+  <%= form.password_field :password %><br><br>
+  <%= form.submit "Submit" %>
+<% end %>
```
Comment (mke2111) [inline app/views/users/new.html.erb:34]: The `form_with` should not be commented out. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_with.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

File: app/views/users/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  <%= form.text_field :email, :value => @user.email %><br>
+  <%= form.label :password, "Edit Password:" %>
+  <%= form.password_field :password %><br>
+  <%= form.submit "Submit" %>
+<% end %>
+<br>
+<%= link_to "<< Back to Users List", root_path %><br/> -->
+<h1>Edit User</h1>
+
+<%= render 'form', user: @user %>
```
Comment (mke2111) [inline app/views/users/edit.html.erb:14]: This file should still contain all the commented-out forms used in the new.html file

([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_edit.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details open="open">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] `form_tag` form is not commented out. Only leave `form_with` uncomment.
- [ ] The previously requested change about where to redirect the user when info is updated hasn't been met yet.



</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [x] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

File: app/views/users/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,34 @@
+<!-- <form action="/users" method="post" accept-charset="UTF-8">
+  <input type="hidden" name="authenticity_token" value="<%= form_authenticity_token %>">
```
Comment (mricanho) [inline app/views/users/new.html.erb]: - [x] This code should be commented

File: app/views/users/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  <%= form.label :email, "Email:" %>
+  <%= form.email_field :email, :value => 'user@domain.com' %><br><br>
+  <%= form.label :password, "Password:" %>
+  <%= form.password_field :password %><br><br>
+  <%= form.submit "Submit" %>
+<% end %>
+<br><br>
+<%= link_to "<< Back to Users List", root_path %><br/> -->
+
+<%= form_tag("/users", method: "post") do %>
```
Comment (mricanho) [inline app/views/users/new.html.erb:25]: - [x] form_tag should be commented entirely

**Review by mricanho — CHANGES_REQUESTED** [review]
# 🔴  Changes Required 🔴  

Hello @LoxiasCam, you did a great job with the forms🔥, just a few changes are required:

- [x]  **form_tag and form_action should be commented enterily** 

You can contact me on slack if you have any questions just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!👨🏽‍💻  🖥️

**Review by mricanho — APPROVED** [review]
# Status: ✔️ Project Approved ✔️ 

Now we are talking @LoxiasCam, good job.

You worked hard in all the instructions that the previous Code Reviewers give you👍🏽 

Well done team

Now you can merge this branch 🚀 

You can contact me on slack if you have any questions, just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!  👨🏽‍💻 🖥️

**Pull request merged**